presto-native-execution/presto_cpp/main/PeriodicMemoryChecker.cpp

```diff
@@ -0,0 +1,198 @@
+  scheduler_->setThreadName("MemoryCheckerThread");
+  scheduler_->addFunction(
+      [&]() {
+        periodicCb();
+        if (config_.mallocMemHeapDumpEnabled) {
+          maybeDumpHeap();
+        }
+        if (config_.systemMemPushbackEnabled &&
+            systemUsedMemoryBytes() > config_.systemMemLimitBytes) {
+          pushbackMemory();
+        }
+      },
+      std::chrono::seconds(config_.MemoryCheckerIntervalSec),
+      "periodic-sys-mem-check",
+      std::chrono::seconds(0));
+  scheduler_->start();
+}
+
+void PeriodicMemoryChecker::stop() {
+  scheduler_->shutdown();
+}
+
+std::string PeriodicMemoryChecker::createHeapDumpFilePath() const {
+  const size_t now = velox::getCurrentTimeMs() / 1000;
+  // Format that can be consumed by strobelight:
+  // <heapDumpFilePrefix>.<pid>.<global_sequence>.i<sequence> =>
+  // prefix.1234.235.i565
+  return fmt::format(
+      "{}/{}.{}.{}.i{}.heap",
+      config_.heapDumpLogDir,
+      config_.heapDumpFilePrefix,
+      getpid(),
+      now,
+      now);
+}
+
+void PeriodicMemoryChecker::maybeDumpHeap() {
+  const auto now = velox::getCurrentTimeMs() / 1000;
+  const auto allocatedSize = mallocBytes();
+  if (allocatedSize >= config_.mallocBytesUsageDumpThreshold &&
+      now - lastHeapDumpAttemptTimestamp_ >= config_.minHeapDumpIntervalSec) {
+    lastHeapDumpAttemptTimestamp_ = now;
+    LOG(INFO) << fmt::format(
+        "Memory usage allocated via malloc exceeded threshold of {}, current "
+        "allocation: {}",
+        velox::succinctBytes(config_.mallocBytesUsageDumpThreshold),
+        velox::succinctBytes(allocatedSize));
+
+    const auto minMemUsageDumped = dumpFilesByHeapMemUsageMinPq_.empty()
+        ? 0
+        : dumpFilesByHeapMemUsageMinPq_.top().mallocUsedBytes;
+    if (dumpFilesByHeapMemUsageMinPq_.size() == config_.maxHeapDumpFiles &&
+        allocatedSize <= minMemUsageDumped) {
+      LOG(INFO) << fmt::format(
+          "Heap profile not dumped as current usage {} is below the "
+          "minimum usage dumped {} and we already have {} files in rotation",
+          velox::succinctBytes(allocatedSize),
+          velox::succinctBytes(minMemUsageDumped),
+          config_.maxHeapDumpFiles);
+      return;
+    }
+    const auto filePath = createHeapDumpFilePath();
+    if (!heapDumpCb(filePath)) {
+      LOG(ERROR) << "Error dumping Heap profile";
+      return;
+    }
+    LOG(INFO) << fmt::format(
+        "Heap profile with usage {} dumped to {}",
+        velox::succinctBytes(allocatedSize),
+        filePath);
+    dumpFilesByHeapMemUsageMinPq_.push({allocatedSize, filePath});
+    if (dumpFilesByHeapMemUsageMinPq_.size() > config_.maxHeapDumpFiles) {
+      auto& evicted = dumpFilesByHeapMemUsageMinPq_.top();
+      LOG(INFO) << fmt::format(
+          "Removing Heap profile with lowest usage {} : {}",
+          velox::succinctBytes(evicted.mallocUsedBytes),
+          evicted.filePath);
+      const int ret = remove(evicted.filePath.c_str());
```

Comment:
remove is velox fs call or std call?

Reply:
system call

Reply:
I guess we need to extract this out as an abstract method.